### ISSUE TITLE
Remove unneeded target configuration/ops from codegen tests.

### ISSUE THREAD

**Issue by benvanik** [iree-org/iree#15253]
I was doing some cleanup in #15254 and noticed that tests under iree/compiler/Codegen/ include a lot of IR they don't (or shouldn't) need, particularly around the exact target specification. Some passes may need this but I doubt most do and it adds noise/fragility to the tests. It'd be good to do a scrub and remove/trim the unneeded information.

As a general rule, unless a pass needs the `hal.executable.export` it should only be the `func.func` from the inner module and not the whole variant. For example, `peel_and_vectorize.mlir` has this test:

```mlir
#compilation = #iree_codegen.compilation_info<
    lowering_config = <tile_sizes = [[64, 64, 0], [8, 32, 0], [0, 0, 16], [0, 0, 0]]>,
    translation_info  = <CPUDoubleTilingPeelingExpert>>
#pipeline_layout = #hal.pipeline.layout<push_constants = 0, sets = [
  #hal.descriptor_set.layout<0, bindings = [
    #hal.descriptor_set.binding<0, storage_buffer>,
    #hal.descriptor_set.binding<1, storage_buffer>,
    #hal.descriptor_set.binding<2, storage_buffer>
  ]>
]>
hal.executable private @preset_config_matmul  {
  hal.executable.variant @system_elf_x86_64 target(<"llvm-cpu", "system-elf-x86_64">) {
    hal.executable.export @no_peel_static_matmul layout(#pipeline_layout) {
    ^bb0(%arg0: !hal.device, %arg1: index, %arg2 : index, %arg3 : index):
      %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root %arg1, %arg2, %arg3
      hal.return %x, %y, %z : index, index, index
    }
    builtin.module {
      func.func @no_peel_static_matmul() {
        %cst = arith.constant 0.000000e+00 : f32
        %lhs_binding = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer)
            : !flow.dispatch.tensor<readonly:tensor<128x64xf32>>
        %rhs_binding = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer)
            : !flow.dispatch.tensor<readonly:tensor<64x512xf32>>
        %result_binding = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer)
            : !flow.dispatch.tensor<writeonly:tensor<128x512xf32>>
        %lhs = flow.dispatch.tensor.load %lhs_binding, offsets = [0, 0], sizes = [128, 64], strides = [1, 1]
            : !flow.dispatch.tensor<readonly:tensor<128x64xf32>> -> tensor<128x64xf32>
        %rhs = flow.dispatch.tensor.load %rhs_binding, offsets = [0, 0], sizes = [64, 512], strides = [1, 1]
            : !flow.dispatch.tensor<readonly:tensor<64x512xf32>> -> tensor<64x512xf32>
        %init = tensor.empty() : tensor<128x512xf32>
        %fill = linalg.fill ins(%cst : f32) outs(%init : tensor<128x512xf32>) -> tensor<128x512xf32>
        %gemm = linalg.matmul {compilation_info = #compilation}
            ins(%lhs, %rhs : tensor<128x64xf32>, tensor<64x512xf32>)
            outs(%fill : tensor<128x512xf32>) -> tensor<128x512xf32>
        flow.dispatch.tensor.store %gemm, %result_binding, offsets = [0, 0], sizes = [128, 512], strides = [1, 1]
            : tensor<128x512xf32> -> !flow.dispatch.tensor<writeonly:tensor<128x512xf32>>
        return
      }
    }
  }
}
```

When it could be:
```mlir
#compilation = #iree_codegen.compilation_info<
    lowering_config = <tile_sizes = [[64, 64, 0], [8, 32, 0], [0, 0, 16], [0, 0, 0]]>,
    translation_info  = <CPUDoubleTilingPeelingExpert>>
func.func @no_peel_static_matmul() {
  %cst = arith.constant 0.000000e+00 : f32
  %lhs_binding = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer)
      : !flow.dispatch.tensor<readonly:tensor<128x64xf32>>
  %rhs_binding = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer)
      : !flow.dispatch.tensor<readonly:tensor<64x512xf32>>
  %result_binding = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer)
      : !flow.dispatch.tensor<writeonly:tensor<128x512xf32>>
  %lhs = flow.dispatch.tensor.load %lhs_binding, offsets = [0, 0], sizes = [128, 64], strides = [1, 1]
      : !flow.dispatch.tensor<readonly:tensor<128x64xf32>> -> tensor<128x64xf32>
  %rhs = flow.dispatch.tensor.load %rhs_binding, offsets = [0, 0], sizes = [64, 512], strides = [1, 1]
      : !flow.dispatch.tensor<readonly:tensor<64x512xf32>> -> tensor<64x512xf32>
  %init = tensor.empty() : tensor<128x512xf32>
  %fill = linalg.fill ins(%cst : f32) outs(%init : tensor<128x512xf32>) -> tensor<128x512xf32>
  %gemm = linalg.matmul {compilation_info = #compilation}
      ins(%lhs, %rhs : tensor<128x64xf32>, tensor<64x512xf32>)
      outs(%fill : tensor<128x512xf32>) -> tensor<128x512xf32>
  flow.dispatch.tensor.store %gemm, %result_binding, offsets = [0, 0], sizes = [128, 512], strides = [1, 1]
      : tensor<128x512xf32> -> !flow.dispatch.tensor<writeonly:tensor<128x512xf32>>
  return
}
```

and probably much less as instead of hal interface bindings it could just use args/results as it doesn't need the extra goo:
```mlir
#compilation = #iree_codegen.compilation_info<
    lowering_config = <tile_sizes = [[64, 64, 0], [8, 32, 0], [0, 0, 16], [0, 0, 0]]>,
    translation_info  = <CPUDoubleTilingPeelingExpert>>
func.func @no_peel_static_matmul(%lhs: tensor<128x64xf32>, %rhs: tensor<64x512xf32>) -> tensor<128x512xf32> {
  %cst = arith.constant 0.000000e+00 : f32
  %init = tensor.empty() : tensor<128x512xf32>
  %fill = linalg.fill ins(%cst : f32) outs(%init : tensor<128x512xf32>) -> tensor<128x512xf32>
  %gemm = linalg.matmul {compilation_info = #compilation}
      ins(%lhs, %rhs : tensor<128x64xf32>, tensor<64x512xf32>)
      outs(%fill : tensor<128x512xf32>) -> tensor<128x512xf32>
  return %gemm : tensor<128x512xf32>
}
```

Now it's a near-upstreamable test that doesn't rely on the entire context of the program (that isn't important here) and is much easier to read/modify/maintain.

Setting some guidelines for new tests being as minimal/vanilla as possible and slowly cleaning up old tests as they are refreshed/people have cycles would be good for project health. As it is now making any changes to the HAL dialects requires updating 500-1000 codegen tests that are each a little different in ways that aren't relevant to the tests :(

**Comment by ScottTodd** [comment]
This has been bugging me too. Complex tests that haven't been reduced down to a minimal snippet are difficult to understand (both to use as a reference and to update when needed).

**Comment by MaheshRavishankar** [comment]
Sorry, I probably was lazy when I started and then things got off-rails. This was also before we could attach a `hal.executable.target` to any parent op.

**Comment by benvanik** [comment]
no worries :) mostly just wanted it tracked as a cleanup task so we could start to whittle away at it

**Comment by ScottTodd** [comment]
Could we update existing test files mechanically? The IR example in the issue description makes that seem possible, at least for some subset of the 100+ test files implicated here.

We have a bunch of tools available:

* regex find/replace
* (new) MLIR passes
* iree-reduce?

**Comment by benvanik** [comment]
worth a shot - there's probably some passes that need to be migrated to use the discovery mechanism and that'd be a good way to find them!

**Comment by ScottTodd** [comment]
I'd like to help break ground on this, if we can work out a few examples to generalize from.

Can the example test actually be trimmed down? It uses this pass pipeline: `--pass-pipeline='builtin.module(hal.executable(hal.executable.variant(iree-codegen-materialize-user-configs, iree-llvmcpu-lower-executable-target)))'`, and both of those inner passes operate on executables, not inner functions:
https://github.com/openxla/iree/blob/c4be76f6ba2e8c64c5dcf4750442a3299b016ce9/compiler/src/iree/compiler/Codegen/Common/MaterializeUserConfigs.cpp#L68-L72 https://github.com/openxla/iree/blob/c4be76f6ba2e8c64c5dcf4750442a3299b016ce9/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L156-L158

> there's probably some passes that need to be migrated to use the discovery mechanism

What's that? I know we have some older passes declared in C++ instead of tablegen... is that different?

**Comment by benvanik** [comment]
Yeah that's what we'll need to fix (where possible). If something needs the exports then it needs the variant (but the variants could often be simplified - don't need workgroup count regions and other stuff if it just wants some attr on the export).

Most things, though, shouldn't need the export. If they do we could instead have a pass that runs at the head of the pipeline and annotates the target func.func with attributes. Then passes could just run on the builtin ModuleOp and check the parent function for its information.

I suspect most, though, are just looking for target configuration and such. `IREE::HAL::ExecutableTargetAttr::lookup(anyOp);` will return the executable target information, so this will work:
```mlir
builtin.module {
  func.func @dispatch() {
    // passes running in here using IREE::HAL::ExecutableTargetAttr::lookup will find the attr below
  } attributes {hal.executable.target: #....}
}
```

**Comment by qedawkins** [comment]
> Yeah that's what we'll need to fix (where possible). If something needs the exports then it needs the variant (but the variants could often be simplified - don't need workgroup count regions and other stuff if it just wants some attr on the export).

This would be really nice to start cleaning up. The more passes that are nested on the variant, the harder it is to break up codegen across entry points (i.e. this comment https://github.com/openxla/iree/blob/c4be76f6ba2e8c64c5dcf4750442a3299b016ce9/compiler/src/iree/compiler/Codegen/Common/MaterializeUserConfigs.cpp#L131). We should be able to choose completely different strategies for different entry points, and for the most part workgroup count is set fairly early on (`TileAndDistributeToWorkgroups`). Similarly, the late stage lowering passes (e.g. -to-spirv + serialization) are all shared across strategies, so it's the strategy specific passes in the middle that need attention.

**Comment by benvanik** [comment]
^ +1! good point about multi-targeting/dynamic dispatch

**Comment by MaheshRavishankar** [comment]
Yeah cleaning that up would be nice. There are two parts to this though.

1) some of not most of those tests only need the target information. Those are used to set tile sizes. The target information could be set anywhere. Those should be "easier"
2) the tile and distribute pass needs to run on executable variant cause they set the workgroup count calculation. That is a bit more involved.
3) the `*-lower-executable-target` is basically the main pass that implements the dynamic pass pipeline to switch between pipelines which implement the tiling and vectorization passes. Since the tile and distribute is part of that this pass runs on the hal.executable.variant . 

The way this is setup, the `*-lower-executabke-target` pass implements the high level "optimizations" like tile and distribute. Then what is remaining is  just a "dumb" lowering to executable code.

To start with maybe all the `materialize-*` passes can drop all the boiler plate and just use the function with the hal target information jus added to the surrounding function.

**Comment by qedawkins** [comment]
> To start with maybe all the `materialize-*` passes can drop all the boiler plate and just use the function with the hal target information jus added to the surrounding function.

If I'm following correctly, I was thinking of doing exactly this as an attempt to deduplicate variant ops before codegen occurs (basically break up `*-lower-executable-target` into `*-materialize-target-strategy` and `*-apply-target-strategy`). This way I could cull duplicate variants as a pass in between. Happy to set up a chat to discuss if needed.

**Comment by MaheshRavishankar** [comment]
> > To start with maybe all the `materialize-*` passes can drop all the boiler plate and just use the function with the hal target information jus added to the surrounding function.
> 
> If I'm following correctly, I was thinking of doing exactly this as an attempt to deduplicate variant ops before codegen occurs (basically break up `*-lower-executable-target` into `*-materialize-target-strategy` and `*-apply-target-strategy`). This way I could cull duplicate variants as a pass in between. Happy to set up a chat to discuss if needed.

So IIUc the `init*Config` for each backend moves into its own pass? If so +1. That's a good idea. Might help here .

**Comment by MaheshRavishankar** [comment]
Actually that split will help quote a bit. We should also put the TranslationInfo on the function instead of the entry point, then we can remove more boiler plate and will also lead to some other things that can help with specialization. I was looking into it, but didn't have the time to go down that path . We can chat more tomorrow.
